### PR TITLE
MODORDERS-1117. Edited location is not displayed in related PO line after receiving piece from full screen form

### PR DESCRIPTION
**Pull request by SerhiiNosko**
## Purpose

https://folio-org.atlassian.net/browse/MODORDERS-1117

## Approach

The intent of this story to make the similar behavior for updating related PO Line for both Full Screen and Quick Receiving views. Currently Quick Receiving view works in correct way: when clicking Create new holding then PO Line Location is updated with new holding. We need to make the same logic for Full Screen Receiving view.

- Do not to call to retrieve pieces again before updating po line
- Use the same logic to update PO Line details(locations, estimated price) in Full Screen Receiving View as in Quick Receive
- Logic to update Po Line details requires pieces from storage to compare with processed ones, so added them in newly created holder
- Use batch endpoint to update po lines instead of one by one

### CODE REVIEW

**Comment by SerhiiNosko** [comment]
Karate tests related with receiving passed
karate_30

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 1]
```diff
@@ -267,29 +272,45 @@ protected Future<Map<String, List<Piece>>> updateOrderAndPoLinesStatus(Map<Strin
...
-        List<Future<String>> futures = new ArrayList<>();
-        for (PoLine poLine : poLines) {
-          if (poLine.getReceiptStatus() == CANCELLED || poLine.getReceiptStatus() == ONGOING) {
-            continue;
-          }
-          List<Piece> successfullyProcessedPieces = getSuccessfullyProcessedPieces(poLine.getId(), piecesGroupedByPoLine);
-          futures.add(calculatePoLineReceiptStatus(poLine, successfullyProcessedPieces, requestContext)
-            .compose(status -> purchaseOrderLineService.updatePoLineReceiptStatus(poLine, status, requestContext)));
+      // Calculate expected status for each PO Line and update with new one if required
+      // Skip status update if PO line status is Ongoing
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:276]: Need to update comment '..or Cancelled"

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 2]
```diff
@@ -267,29 +272,45 @@ protected Future<Map<String, List<Piece>>> updateOrderAndPoLinesStatus(Map<Strin
...
+      List<Future<PoLine>> poLinesToUpdate = new ArrayList<>();
+      for (PoLine poLine : poLines) {
+        if (poLine.getReceiptStatus() == CANCELLED || poLine.getReceiptStatus() == ONGOING) {
+          continue;
+        }
+
+        List<Piece> successfullyProcessedPieces = getSuccessfullyProcessedPieces(poLine.getId(), piecesGroupedByPoLine);
+        if (CollectionUtils.isEmpty(successfullyProcessedPieces)) {
+          logger.info("updateRelatedPoLineDetails:: No pieces processed - nothing to update for PO Line '{}'", poLine.getId());
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:285]: need to update
```
logger.info("updateOrderAndPoLinesStatus:: No pieces processed - nothing to update for PO Line '{}'", poLine.getId());
```

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 3]
```diff
@@ -267,29 +272,45 @@ protected Future<Map<String, List<Piece>>> updateOrderAndPoLinesStatus(Map<Strin
...
-          logger.debug("{} out of {} PO Line(s) updated with new status", updatedPoLines.size(), piecesGroupedByPoLine.size());
+        Future<PoLine> poLineToUpdateFuture = getPiecesByPoLine(poLine.getId(), requestContext)
+          .map(pieces -> updateRelatedPoLineDetails(poLine, piecesFromStorage.get(poLine.getId()), pieces, successfullyProcessedPieces));
 
-          List<PoLine> successPoLines = StreamEx.of(poLines)
-            .filter(line -> updatedPoLines.contains(line.getId()))
-            .toList();
-          updateOrderStatus.accept(successPoLines);
-          return null;
+        poLinesToUpdate.add(poLineToUpdateFuture);
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:292]: at first sight, it is confusing we updated poLine and naming to poLineToUpdate. Maybe need to specify what need to update . poLineToUpdateLocation.

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 4]
```diff
@@ -267,29 +272,45 @@ protected Future<Map<String, List<Piece>>> updateOrderAndPoLinesStatus(Map<Strin
...
+        .map(updatedPoLines -> {
+          updatedPoLines.forEach(v -> logger.info("updateOrderAndPoLinesStatus:: updating PoLine: {}", JsonObject.mapFrom(v).getJsonArray("locations").encodePrettily()));
+
+          return purchaseOrderLineService.saveOrderLines(updatedPoLines, requestContext)
+            .map(voidResult -> {
+              logger.info("updateOrderAndPoLinesStatus:: {} out of {} PO Line(s) updated with new status in batch", poLines.size(), piecesGroupedByPoLine.size());
+
+              updateOrderStatus.accept(poLines);
+              return null;
+            });
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:310]: it is already big method, I suggest to extract to method

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 5]
```diff
@@ -267,29 +272,45 @@ protected Future<Map<String, List<Piece>>> updateOrderAndPoLinesStatus(Map<Strin
...
-          List<PoLine> successPoLines = StreamEx.of(poLines)
-            .filter(line -> updatedPoLines.contains(line.getId()))
-            .toList();
-          updateOrderStatus.accept(successPoLines);
-          return null;
+        poLinesToUpdate.add(poLineToUpdateFuture);
+      }
+
+      if (CollectionUtils.isEmpty(poLinesToUpdate)) {
+        logger.info("updateRelatedPoLineDetails:: no po lines found for update, returning");
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:296]: returning is not clear what is being return, instead we can remove it
```
 logger.info("updateOrderAndPoLinesStatus:: No po lines found to update");
```

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 6]
```diff
@@ -308,52 +329,82 @@ private List<Piece> getSuccessfullyProcessedPieces(String poLineId, Map<String,
...
+   * @param piecesByPoLine               all piece records obtained by PO line id
+   * @param piecesSuccessfullyProcessed  pieces that were successfully processed
+   * @param poLine                       PO Line record representation from storage
    * @return completable future holding calculated PO Line's receipt status
    */
-  private Future<ReceiptStatus> calculatePoLineReceiptStatus(List<Piece> byPoLine, PoLine poLine, List<Piece> pieces) {
-    long expectedPiecesQuantity = byPoLine.stream()
+  private ReceiptStatus calculatePoLineReceiptStatus(List<Piece> piecesByPoLine, List<Piece> piecesSuccessfullyProcessed, PoLine poLine) {
+    logger.info("calculatePoLineReceiptStatus:: piecesByPoLine: {}", piecesByPoLine.size());
+    logger.info("calculatePoLineReceiptStatus:: pieces: {}", piecesSuccessfullyProcessed.size());
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:382]: lets add meaningful messages, so it can be usefuly to understand flow

Comment (azizbekxm) [reply src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:382]: as suggestion, 
```
logger.info("calculatePoLineReceiptStatus:: Trying to calculate poLineReceiptStatus with id '{}' using '{}' piecesByPoLine and '{}' piece(s) and", poLine.getId(), piecesByPoLine.size(), piecesSuccessfullyProcessed.size())
```

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 7]
```diff
@@ -308,52 +329,82 @@ private List<Piece> getSuccessfullyProcessedPieces(String poLineId, Map<String,
...
+    logger.info("calculatePoLineReceiptStatus:: FULLY_RECEIVED: {}", !poLine.getCheckinItems().equals(Boolean.TRUE) && expectedPiecesQuantity == 0);
     if (!poLine.getCheckinItems().equals(Boolean.TRUE) && expectedPiecesQuantity == 0) {
-      return Future.succeededFuture(FULLY_RECEIVED);
+      return FULLY_RECEIVED;
     }
     // Partially Received: In case there is at least one successfully received piece
-    if (StreamEx.of(pieces).anyMatch(piece -> RECEIVED_STATUSES.contains(piece.getReceivingStatus()))) {
-      return Future.succeededFuture(PARTIALLY_RECEIVED);
+    logger.info("calculatePoLineReceiptStatus:: PARTIALLY_RECEIVED: {}", StreamEx.of(piecesSuccessfullyProcessed).anyMatch(piece -> RECEIVED_STATUSES.contains(piece.getReceivingStatus())));
+    if (StreamEx.of(piecesSuccessfullyProcessed).anyMatch(piece -> RECEIVED_STATUSES.contains(piece.getReceivingStatus()))) {
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:397]: seems duplicate code, better to extract method in helperUtils or somewhere else

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 8]
```diff
@@ -308,52 +329,82 @@ private List<Piece> getSuccessfullyProcessedPieces(String poLineId, Map<String,
...
     }
     // Pieces were rolled-back to Expected. In this case we have to check if
     // there is any Received piece in the storage
-    long receivedQty = byPoLine.stream()
+    long receivedQty = piecesByPoLine.stream()
       .filter(piece -> RECEIVED_STATUSES.contains(piece.getReceivingStatus()))
       .count();
-    return Future.succeededFuture(receivedQty == 0 ? AWAITING_RECEIPT : PARTIALLY_RECEIVED);
+    logger.info("calculatePoLineReceiptStatus:: receivedQty: {}", receivedQty);
+    logger.info("calculatePoLineReceiptStatus:: AWAITING_RECEIPT or PARTIALLY_RECEIVED: {}", receivedQty == 0 ? AWAITING_RECEIPT : PARTIALLY_RECEIVED);
```
Comment (azizbekxm) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:406]: as suggestion earlier, log messages are very hard to understand its purpose. better to add meaningful message and optimize if possible

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java (hunk 3 above)
Comment (BKadirkhodjaev) [reply src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:292]: > at first sight, it is confusing we updated poLine and naming to poLineToUpdate. Maybe need to specify what need to update . poLineToUpdateLocation.

It updates more than just location though

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java (hunk 6 above)
Comment (BKadirkhodjaev) [reply src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:382]: These were my diagnostic logs, will either remove or modify them

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java [hunk 9]
```diff
@@ -267,29 +272,47 @@ protected Future<Map<String, List<Piece>>> updateOrderAndPoLinesStatus(Map<Strin
...
 
-          List<PoLine> successPoLines = StreamEx.of(poLines)
-            .filter(line -> updatedPoLines.contains(line.getId()))
-            .toList();
-          updateOrderStatus.accept(successPoLines);
-          return null;
-        });
-      })
-      .map(ok -> piecesGroupedByPoLine);
+        Future<PoLine> poLineToUpdateFuture = getPiecesByPoLine(poLine.getId(), requestContext)
```
Comment (BKadirkhodjaev) [inline src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:290]: Necessary for Receipt Status to be correctly determined

**Review by BKadirkhodjaev — APPROVED** [review]

**Review by azizbekxm — APPROVED** [review]

File: src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java (hunk 9 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:290]: can we use piecesFromStorage here that we already have in holder?

Comment (BKadirkhodjaev) [reply src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java:290]: > can we use piecesFromStorage here that we already have in holder?

From my debugging sessions, these pieces are not every piece for a given POL, if we use this one we will get inconsistent Receipt Status as it requires all pieces to be included in the calculation

**Pull request merged**